torch_geometric/loader/node_loader.py

```diff
@@ -0,0 +1,110 @@
+from typing import Any, Callable, Dict, Iterator, Optional, Tuple, Union
+
+import torch
+
+from torch_geometric.data import Data, HeteroData
+from torch_geometric.data.feature_store import FeatureStore
+from torch_geometric.data.graph_store import GraphStore
+from torch_geometric.loader.base import DataLoaderIterator
+from torch_geometric.loader.utils import (
+    filter_custom_store,
+    filter_data,
+    filter_hetero_data,
+    get_input_nodes,
+)
+from torch_geometric.sampler.base import (
+    BaseSampler,
+    HeteroSamplerOutput,
+    NodeSamplerInput,
+    SamplerOutput,
+)
+from torch_geometric.typing import InputNodes
+
+
+class NodeLoader(torch.utils.data.DataLoader):
+    r"""A data loader that performs neighborhood sampling from node information."""
+    def __init__(
+        self,
+        data: Union[Data, HeteroData, Tuple[FeatureStore, GraphStore]],
```

Comment:
Random question I'll ask here: Do we imagine supporting all three types in the Future? If so:
- should we define a type for `Tuple[FeatureStore, GraphStore]]`, its used in many places
- do you think its possible to make `Data` provide the same interface as `HeteroData` (i.e a dictionary with edge_type) so we can reduce a lot of special case handling?

Reply:
> should we define a type for Tuple[FeatureStore, GraphStore]], its used in many places

Good point; let me handle this in a follow-up PR (since this one is already a little large).

> do you think its possible to make Data provide the same interface as HeteroData (i.e a dictionary with edge_type) so we can reduce a lot of special case handling?

Another great point, I've been thinking about this a bit as well. I think that would indeed reduce a lot of special case handling; will take a closer look in a follow-up (after all this main refactoring is complete).

Reply:
Cool sounds good - Do we have a roadmap issue for this? If so maybe I can help out with a couple of these issue too. Thanks for your thoughts!

Reply:
I haven't had the time to set up a roadmap issue yet, but will handle this tomorrow and tag you. Thanks for helping out :)